presto-cli/src/main/java/com/facebook/presto/cli/QueryPreprocessor.java

```diff
@@ -0,0 +1,164 @@
+    private static final String PREPROCESSING_QUERY_MESSAGE = "Preprocessing query";
+
+    private QueryPreprocessor() {}
+
+    public static String preprocessQuery(Optional<String> catalog, Optional<String> schema, String query)
+            throws QueryPreprocessorException
+    {
+        Duration timeout = DEFAULT_PREPROCESSOR_TIMEOUT;
+        String timeoutEnvironment = System.getenv(ENV_PREPROCESSOR_TIMEOUT);
+        if (emptyToNull(timeoutEnvironment) != null) {
+            timeout = Duration.valueOf(timeoutEnvironment.trim());
+        }
+
+        String preprocessorCommand = System.getenv(ENV_PREPROCESSOR);
+        if (emptyToNull(preprocessorCommand) == null) {
+            return query;
+        }
+        return preprocessQuery(catalog, schema, query, ImmutableList.of("/bin/sh", "-c", preprocessorCommand.trim()), timeout);
+    }
+
+    public static String preprocessQuery(Optional<String> catalog, Optional<String> schema, String query, List<String> preprocessorCommand, Duration timeout)
+            throws QueryPreprocessorException
+    {
+        Thread clientThread = Thread.currentThread();
+        SignalHandler oldHandler = Signal.handle(SIGINT, signal -> clientThread.interrupt());
+        try {
+            if (REAL_TERMINAL) {
+                System.out.print(PREPROCESSING_QUERY_MESSAGE);
+                System.out.flush();
+            }
+            return preprocessQueryInternal(catalog, schema, query, preprocessorCommand, timeout);
+        }
+        finally {
+            if (REAL_TERMINAL) {
+                System.out.print("\r" + Strings.repeat(" ", PREPROCESSING_QUERY_MESSAGE.length()) + "\r");
+                System.out.flush();
+            }
+            Signal.handle(SIGINT, oldHandler);
+            Thread.interrupted(); // clear interrupt status
+        }
+    }
+
+    private static String preprocessQueryInternal(Optional<String> catalog, Optional<String> schema, String query, List<String> preprocessorCommand, Duration timeout)
+            throws QueryPreprocessorException
+    {
+        // execute the process in a child thread so we can better handle interruption and timeouts
+        AtomicReference<Process> processReference = new AtomicReference<>();
+
+        FutureTask<String> task = new FutureTask<>(() -> {
+            String result;
+            int exitCode;
+            try {
+                ProcessBuilder processBuilder = new ProcessBuilder(preprocessorCommand);
+                processBuilder.redirectErrorStream(true);
+                processBuilder.environment().put(ENV_PRESTO_CATALOG, catalog.orElse(""));
+                processBuilder.environment().put(ENV_PRESTO_SCHEMA, schema.orElse(""));
+
+                Process process = processBuilder.start();
+                try {
+                    // write query to process standard out
+                    OutputStream outputStream = process.getOutputStream();
+                    outputStream.write(query.getBytes(UTF_8));
```

Comment:
Very good point.  I'll check in this version and then write a new PR to move the writer to another thread.